File: programmers/K번째수.py
# 최근 풀이법
def solution(array, commands):
    return [sorted(array[c[0]-1:c[1]])[c[2]-1] for c in commands]

# 옛날 풀이법
def solution(array, commands):
    answer = []
    for c in commands:
        answer.append(sorted(array[c[0]-1:c[1]])[c[2]-1])
    return answer

# 다른 사람의 풀이법
def solution(array, commands):
    return list(map(lambda x:sorted(array[x[0]-1:x[1]])[x[2]-1], commands))

File: programmers/고득점 Kit/sort/K번째수.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

/**
 * @brief sort -> O(k log k)
 * @param array 
 * @param commands 
 * @return vector<int> 
 */
vector<int> mySolution(vector<int> array, vector<vector<int>> commands) {
    vector<int> answer;
    
    for (const auto& cmd: commands) {
        vector<int> temp(array.begin() + (cmd[0] - 1), array.begin() + cmd[1]);
        sort(temp.begin(), temp.end());
        answer.push_back(temp[cmd[2] - 1]);
    }
    
    return answer;
}

/**
 * @brief nth_element -> 평균 O(k), 최악 O(k^2)
 * @param array 
 * @param commands 
 * @return vector<int> 
 */
vector<int> gptSolution(vector<int> array, vector<vector<int>> commands) {
    vector<int> answer;

    for (const auto& cmd: commands) {
        vector<int> temp(array.begin() + (cmd[0] - 1), array.begin() + cmd[1]);
        nth_element(temp.begin(), temp.begin() + (cmd[2] - 1), temp.end());
        answer.push_back(temp[cmd[2] - 1]);
    }

    return answer;
}